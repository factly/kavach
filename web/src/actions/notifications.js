import { ADD_NOTIFICATION } from '../constants/notifications';

export const addErrorNotification = (data) => ({
  type: ADD_NOTIFICATION,
  payload: {
    type: 'error',
    title: 'Error',
    message: data,
  },
});

export const addSuccessNotification = (data) => ({
  type: ADD_NOTIFICATION,
  payload: {
    type: 'success',
    title: 'Success',
    message: data,
  },
});
