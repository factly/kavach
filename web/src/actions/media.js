import axios from 'axios';
import {
  ADD_MEDIUM,
  ADD_MEDIA,
  ADD_MEDIA_REQUEST,
  SET_MEDIA_LOADING,
  RESET_MEDIA,
  MEDIA_API,
} from '../constants/media';
import { buildObjectOfItems } from '../utils/objects';

import { addErrorNotification, addSuccessNotification } from './notifications';

export const getMedia = (query) => {
  return (dispatch) => {
    dispatch(loadingMedia());
    return axios
      .get(MEDIA_API, {
        params: query,
      })
      .then((response) => {
        dispatch(addMediaList(response.data.nodes));
        dispatch(
          addMediaRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
        dispatch(stopMediaLoading());
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const getMedium = (id) => {
  return (dispatch) => {
    dispatch(loadingMedia());
    return axios
      .get(MEDIA_API + '/' + id)
      .then((response) => {
        dispatch(getMediumByID(response.data));
        dispatch(stopMediaLoading());
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const addMedium = (data) => {
  return (dispatch) => {
    dispatch(loadingMedia());
    return axios
      .post(MEDIA_API, data)
      .then(() => {
        dispatch(resetMedia());
        dispatch(addSuccessNotification('Media Added'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const updateMedium = (data) => {
  return (dispatch) => {
    dispatch(loadingMedia());
    return axios
      .put(MEDIA_API + '/' + data.id, data)
      .then((response) => {
        dispatch(getMediumByID(response.data));
        dispatch(stopMediaLoading());
        dispatch(addSuccessNotification('Media Updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const deleteMedium = (id) => {
  return (dispatch) => {
    dispatch(loadingMedia());
    return axios
      .delete(MEDIA_API + '/' + id)
      .then(() => {
        dispatch(resetMedia());
        dispatch(addSuccessNotification('Media Deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const addMedia = (media) => {
  return (dispatch) => {
    dispatch(addMediaList(media));
  };
};

export const loadingMedia = () => ({
  type: SET_MEDIA_LOADING,
  payload: true,
});

export const stopMediaLoading = () => ({
  type: SET_MEDIA_LOADING,
  payload: false,
});

export const getMediumByID = (data) => ({
  type: ADD_MEDIUM,
  payload: data,
});

export const addMediaList = (data) => (dispatch) => {
  dispatch({
    type: ADD_MEDIA,
    payload: buildObjectOfItems(data),
  });
};

export const addMediaRequest = (data) => ({
  type: ADD_MEDIA_REQUEST,
  payload: data,
});

export const resetMedia = () => ({
  type: RESET_MEDIA,
});
