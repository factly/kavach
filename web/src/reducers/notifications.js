import { ADD_NOTIFICATION } from '../constants/notifications';

const initialState = {
  type: null,
  message: null,
  description: null,
};

export default function notificationsReducer(state = initialState, action = {}) {
  if (!action.payload) {
    return state;
  }
  switch (action.type) {
    case ADD_NOTIFICATION:
      return {
        ...state,
        type: action.payload.type,
        message: action.payload.title,
        description: action.payload.message,
      };
    default:
      return state;
  }
}
