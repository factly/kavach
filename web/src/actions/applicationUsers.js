import axios from 'axios';
import {
  ADD_APPLICATION_USER,
  ADD_APPLICATION_USERS,
  ADD_APPLICATION_USERS_REQUEST,
  SET_APPLICATION_USERS_LOADING,
  RESET_APPLICATION_USERS,
  APPLICATION_USERS_API,
  ADD_USER_IDS,
} from '../constants/applicationUser';
import { getIds } from '../utils/objects';
import { addMediaList } from './media';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { loadingUsers, stopUsersLoading } from './users';

export const getApplications = (appID) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .get(
        APPLICATION_USERS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          appID +
          '/users',
      )
      .then((response) => {
        dispatch(
          addApplicationList(
            response.data.users.map((user) => {
              return user;
            }),
          ),
        );
        dispatch(
          addApplicationsRequest({
            data: response.data.users.map((item) => item.id),
            application_id: response.data.application.id,
            total: response.data.total,
          }),
        );
        dispatch(stopApplicationLoading());
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const getApplication = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .get(
        APPLICATION_USERS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          id +
          '/users',
      )
      .then((response) => {
        if (response.data.medium) dispatch(addMediaList([response.data.medium]));
        dispatch(getApplicationByID({ ...response.data, medium: response.data.medium?.id }));
        dispatch(stopApplicationLoading());
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const addApplicationUser = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .post(
        APPLICATION_USERS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          data.application_id +
          '/users',
        data,
      )
      .then(() => {
        dispatch(resetApplications());
        dispatch(addSuccessNotification('Application User Added'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const updateApplication = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .put(
        APPLICATION_USERS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          data.id,
        data,
      )
      .then((response) => {
        if (response.data.medium) dispatch(addMediaList([response.data.medium]));
        dispatch(getApplicationByID({ ...response.data, medium: response.data.medium?.id }));
        dispatch(stopApplicationLoading());
        dispatch(addSuccessNotification('Application Updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const deleteApplication = (id, appID) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .delete(
        APPLICATION_USERS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          id +
          '/users/' +
          appID,
      )
      .then(() => {
        dispatch(resetApplications());
        dispatch(addSuccessNotification('Application User Deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const getApplicationUsers = (appID) => {
  return (dispatch, getState) => {
    dispatch(loadingUsers());
    return axios
      .get(
        APPLICATION_USERS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          appID +
          '/users',
      )
      .then((res) => {
        dispatch(addUserIds(getIds(res.data.users), appID));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopUsersLoading());
      });
  };
};

export const addApplications = (applications) => {
  return (dispatch) => {
    dispatch(
      addMediaList(
        applications
          .filter((application) => application.medium)
          .map((application) => application.medium),
      ),
    );
    dispatch(
      addApplicationList(
        applications.map((application) => {
          return { ...application, medium: application.medium?.id };
        }),
      ),
    );
  };
};

export const loadingApplications = () => ({
  type: SET_APPLICATION_USERS_LOADING,
  payload: true,
});

export const stopApplicationLoading = () => ({
  type: SET_APPLICATION_USERS_LOADING,
  payload: false,
});

export const getApplicationByID = (data) => ({
  type: ADD_APPLICATION_USER,
  payload: data,
});

export const addApplicationList = (data) => ({
  type: ADD_APPLICATION_USERS,
  payload: data,
});

export const addApplicationsRequest = (data) => ({
  type: ADD_APPLICATION_USERS_REQUEST,
  payload: data,
});

const addUserIds = (data, appID) => ({
  type: ADD_USER_IDS,
  payload: {
    id: appID,
    data: data,
  },
});
export const resetApplications = () => ({
  type: RESET_APPLICATION_USERS,
});
