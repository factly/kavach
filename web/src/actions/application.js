import axios from 'axios';
import {
  ADD_APPLICATION,
  ADD_APPLICATIONS,
  ADD_APPLICATIONS_REQUEST,
  SET_APPLICATIONS_LOADING,
  RESET_APPLICATIONS,
  APPLICATIONS_API,
  ADD_SPACE_IDS,
  SET_DEFAULT_APPLICATION_LOADING,
  GET_DEFAULT_APPLICATIONS,
} from '../constants/application';
import { ADD_APPLICATION_IDS, ORGANISATIONS_API } from '../constants/organisations';
import { buildObjectOfItems, deleteKeys, getIds, getValues } from '../utils/objects';
import { addMedia, addMediaList } from './media';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { addSpaces } from './space';
import { addApplicationTokens } from './token';
import { addUsersList } from './users';
export const getApplications = () => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .get(APPLICATIONS_API + '/' + getState().organisations.selected + '/applications')
      .then((response) => {
        dispatch(addApplicationList(response.data));
        dispatch(addApplicationIds(getIds(response.data)));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopApplicationLoading());
      });
  };
};

export const addDefaultApplication = (appID) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .post(
        ORGANISATIONS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          appID +
          '/default',
      )
      .then(() => {
        dispatch(addSuccessNotification('Application Added Successfully'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopApplicationLoading());
      });
  };
};

export const removeDefaultApplication = (appID) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .delete(
        ORGANISATIONS_API +
          '/' +
          getState().organisations.selected +
          '/applications/' +
          appID +
          '/default',
      )
      .then(() => {
        dispatch(addSuccessNotification('Application removed succesfully'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopApplicationLoading());
      });
  };
};

export const getDefaultApplications = () => {
  return (dispatch, getState) => {
    dispatch(setDefaultApplicationsLoading());
    return axios
      .get(ORGANISATIONS_API + '/' + getState().organisations.selected + '/applications/default')
      .then((response) => {
        dispatch(addDefaultApplicationsList(response.data));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopDefaultApplicationLoading());
      });
  };
};

export const getApplication = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .get(APPLICATIONS_API + '/' + getState().organisations.selected + '/applications/' + id)
      .then((response) => {
        if (response.data.medium_id) {
          dispatch(addMedia(response.data.medium));
        }
        deleteKeys([response.data], ['medium']);
        dispatch(addUsersList(response.data.users));
        response.data.users = getIds(response.data.users);
        response.data.spaces = getIds(response.data.spaces)
        response.data.tokens = getIds(response.data.tokens)
        const spaces = getValues([response.data], 'spaces');
        // dispatch(addSpaces(spaces));
        dispatch(addApplication(response.data));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopApplicationLoading());
      });
  };
};

export const createApplication = (data) => {
  data.is_default = false;
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .post(ORGANISATIONS_API + '/' + getState().organisations.selected + '/applications', data)
      .then(() => {
        dispatch(resetApplications());
        dispatch(addSuccessNotification('Application Added'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const updateApplication = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .put(
        ORGANISATIONS_API + '/' + getState().organisations.selected + '/applications/' + data.id,
        data,
      )
      .then(() => {
        dispatch(addSuccessNotification('Application Updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopApplicationLoading());
      });
  };
};

export const deleteApplication = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingApplications());
    return axios
      .delete(ORGANISATIONS_API + '/' + getState().organisations.selected + '/applications/' + id)
      .then(() => {
        dispatch(resetApplications());
        dispatch(addSuccessNotification('Application Deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const addApplications = (applications) => {
  return (dispatch) => {
    dispatch(
      addMediaList(
        applications
          .filter((application) => application.medium)
          .map((application) => application.medium),
      ),
    );
    dispatch(
      addApplicationList(
        applications.map((application) => {
          return { ...application, medium: application.medium?.id };
        }),
      ),
    );
  };
};

export const loadingApplications = () => ({
  type: SET_APPLICATIONS_LOADING,
  payload: true,
});

export const stopApplicationLoading = () => ({
  type: SET_APPLICATIONS_LOADING,
  payload: false,
});

export const getApplicationByID = (data) => ({
  type: ADD_APPLICATION,
  payload: data,
});

export const addApplicationList = (data) => (dispatch) => {
  dispatch(loadingApplications());
  const medium = getValues(data, 'medium');
  dispatch(addMediaList(medium));
  deleteKeys(data, ['medium']);
  const spaces = getValues(data, 'spaces');
  dispatch(addSpaces(spaces));

  data.forEach((application) => {
    application.spaces = getIds(application.spaces);
    application.users = getIds(application.users);
    dispatch(addApplicationTokens(application.id, application.tokens));
    application.tokens = getIds(application.tokens);
  });

  dispatch({
    type: ADD_APPLICATIONS,
    payload: buildObjectOfItems(data),
  });
  dispatch(stopApplicationLoading());
};

export const addApplicationsRequest = (data) => ({
  type: ADD_APPLICATIONS_REQUEST,
  payload: data,
});

export const resetApplications = () => ({
  type: RESET_APPLICATIONS,
});

export const addApplicationIds = (data) => ({
  type: ADD_APPLICATION_IDS,
  payload: data,
});

export const addApplication = (data) => ({
  type: ADD_APPLICATION,
  payload: data,
});

export const addSpaceIDs = (appID, data) => ({
  type: ADD_SPACE_IDS,
  payload: {
    appID: appID,
    data: data,
  },
});

export const setDefaultApplicationsLoading = () => ({
  type: SET_DEFAULT_APPLICATION_LOADING,
  payload: true,
});

export const stopDefaultApplicationLoading = () => ({
  type: SET_DEFAULT_APPLICATION_LOADING,
  payload: false,
});

export const addDefaultApplicationsList = (data) => ({
  type: GET_DEFAULT_APPLICATIONS,
  payload: data,
});
