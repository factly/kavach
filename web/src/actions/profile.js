import axios from 'axios';
import {
  SET_PROFILE_LOADING,
  ADD_PROFILE,
  PROFILE_API,
  ADD_INVITE,
  DELETE_INVITE,
} from '../constants/profile';
import { buildObjectOfItems, deleteKeys, getIds } from '../utils/objects';
import { addApplicationsList, loadingApplications, stopApplicationLoading } from './application';
import { addMedia } from './media';
import { addErrorNotification, addSuccessNotification } from './notifications';
import {
  addOrganisationsList,
  getOrganisations,
  loadingOrganisations,
  stopOrganisationsLoading,
} from './organisations';
import { addSpaces, stopLoadingSpaces, loadingSpaces } from './space';
import { addUsersList, loadingUsers, stopUsersLoading } from './users';

export const getUserProfile = () => {
  return (dispatch, getState) => {
    dispatch(loadingProfile());
    return axios
      .get(PROFILE_API)
      .then((response) => {
        dispatch(getProfile(response.data));
        dispatch(stopProfileLoading());
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};
export const updateProfile = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingProfile());
    return axios
      .put(PROFILE_API, data)
      .then((response) => {
        dispatch(getProfile(response.data));
        dispatch(stopProfileLoading());
        dispatch(addSuccessNotification('Profile Updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const getInvitation = () => {
  return (dispatch, getState) => {
    dispatch(loadingProfile);
    return axios
      .get(PROFILE_API + '/invite')
      .then((response) => {
        dispatch(getInvite(response.data));
        dispatch(stopProfileLoading());
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const deleteInvitation = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingProfile);
    return axios
      .delete(PROFILE_API + '/invite/' + id)
      .then(() => {
        dispatch(deleteInvite(id));
        dispatch(stopProfileLoading());
        dispatch(addSuccessNotification('Request declined successfully'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const acceptInvitation = (id, data) => {
  return (dispatch, getState) => {
    dispatch(loadingProfile);
    return axios
      .put(PROFILE_API + '/invite/' + id, data)
      .then(() => {
        dispatch(deleteInvite(id));
        dispatch(getOrganisations());
        dispatch(addSuccessNotification('Request accepted successfully'));
        dispatch(stopProfileLoading());
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const getProfileDetails = () => {
  return (dispatch, getState) => {
    dispatch(loadingProfile);
    dispatch(loadingUsers());
    dispatch(loadingSpaces());
    dispatch(loadingOrganisations());
    dispatch(loadingApplications());
    return axios
      .get(PROFILE_API + '/details')
      .then((response) => {
        response.data.spaces.map((space) => {
          space.user_ids = getIds(space.users);
          dispatch(addUsersList(buildObjectOfItems(space.users)));
          if (space.logo_id !== null) {
            dispatch(addMedia(space.logo));
          }
          deleteKeys([space], ['logo']);
          if (space.logo_mobile_id !== null) {
            dispatch(addMedia(space.logo_mobile));
          }
          deleteKeys([space], ['logo_mobile']);
          if (space.fav_icon_id !== null) {
            dispatch(addMedia(space.fav_icon));
          }
          deleteKeys([space], ['fav_icon']);
          if (space.mobile_icon_id !== null) {
            dispatch(addMedia(space.mobile_icon));
          }
          deleteKeys([space], ['mobile_icon']);
          return null;
        });
        deleteKeys(response.data.spaces, ['users']);
        dispatch(addSpaces(buildObjectOfItems(response.data.spaces)));
        response.data.applications.map((application) => {
          if (application.medium_id !== null) {
            dispatch(addMedia(application.medium));
          }
          deleteKeys([application], ['medium']);
          application.user_ids = getIds(application.users);
          application.space_ids = [];
          response.data.spaces.map((space) => {
            if (space.application_id === application.id) {
              application.space_ids.push(space.id);
            }
            return null;
          });
          dispatch(addUsersList(buildObjectOfItems(application.users)));
          return null;
        });
        deleteKeys(response.data.applications, ['users']);
        dispatch(addApplicationsList(buildObjectOfItems(response.data.applications)));
        response.data.organisations.map((organisation) => {
          if (organisation.featured_medium_id !== null) {
            dispatch(addMedia(organisation.medium));
          }
          deleteKeys([organisation], ['medium']);
          organisation.application_ids = [];
          response.data.applications.map((application) => {
            if (application.organisation_id === organisation.id) {
              organisation.application_ids.push(application.id);
            }
            return null;
          });
          return null;
        });
        deleteKeys(response.data.organisations, ['users']);
        dispatch(addOrganisationsList(response.data.organisations));
        response.data.organisation_ids = getIds(response.data.organisations);
        deleteKeys([response.data], ['applications', 'spaces', 'organisations']);
        if (response.data.featured_medium_id !== null) {
          dispatch(addMedia(response.data.medium));
        }
        deleteKeys([response.data], ['medium']);
        dispatch(getProfile(response.data));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopProfileLoading());
        dispatch(stopUsersLoading());
        dispatch(stopLoadingSpaces());
        dispatch(stopOrganisationsLoading());
        dispatch(stopApplicationLoading());
      });
  };
};

export const loadingProfile = () => ({
  type: SET_PROFILE_LOADING,
  payload: true,
});

export const stopProfileLoading = () => ({
  type: SET_PROFILE_LOADING,
  payload: false,
});

export const getProfile = (data) => ({
  type: ADD_PROFILE,
  payload: data,
});

export const getInvite = (data) => ({
  type: ADD_INVITE,
  payload: data,
});

export const deleteInvite = (data) => ({
  type: DELETE_INVITE,
  payload: data,
});
