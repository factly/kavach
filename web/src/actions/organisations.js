import axios from 'axios';
import {
  ADD_ORGANISATION,
  ADD_ORGANISATIONS,
  SET_ORGANISATIONS_LOADING,
  RESET_ORGANISATIONS,
  ORGANISATIONS_API,
  SET_SELECTED_ORGANISATION,
} from '../constants/organisations';
import { ADD_ORGANISATION_IDS } from '../constants/profile';
import { buildObjectOfItems, deleteKeys, getIds, getValues } from '../utils/objects';
import { addApplicationList } from './application';
import { addMedia, addMediaList } from './media';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { addUsersList } from './users';

export const getOrganisations = () => {
  return (dispatch, getState) => {
    dispatch(loadingOrganisations());
    return axios
      .get(ORGANISATIONS_API + '/my')
      .then((response) => {
        dispatch(addOrganisationsList(response.data));
        dispatch(addOrganisationIds(getIds(response.data)));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopOrganisationsLoading());
      });
  };
};

export const getOrganisation = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingOrganisations());
    return axios
      .get(ORGANISATIONS_API + '/' + id)
      .then((response) => {
        if (response.featured_medium_id) {
          addMedia(response.data.medium);
        }
        let users = [];
        response.data.roles = {};
        response.data.organisation_users.map((item) => {
          users.push(item.user);
          response.data.roles[item.user.id] = item.role;
          return null;
        });
        response.data.role = response.data.permission.role;
        deleteKeys([response.data], ['permission', 'organisation_users']);
        response.data.applications = getIds(response.data.applications);
        response.data.users = getIds(users);
        dispatch(getOrganisationByID(response.data));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(()=>{
        dispatch(stopOrganisationsLoading());
      })
      ;
  };
};

export const addOrganisation = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingOrganisations());
    return axios
      .post(ORGANISATIONS_API, data)
      .then((response) => {
        dispatch(getOrganisationByID(response.data));
        dispatch(setSelectedOrganisation(response.data.id));
        dispatch(addSuccessNotification('Organisation added'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      })
      .finally(() => {
        dispatch(stopOrganisationsLoading());
      });
  };
};

export const updateOrganisation = (data) => {
  return (dispatch, getState) => {
    dispatch(loadingOrganisations());
    return axios
      .put(ORGANISATIONS_API + '/' + data.id, data)
      .then((response) => {
        dispatch(getOrganisationByID(response.data));
        dispatch(stopOrganisationsLoading());
        dispatch(addSuccessNotification('Organisation Updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const deleteOrganisation = (id) => {
  return (dispatch, getState) => {
    dispatch(loadingOrganisations());
    return axios
      .delete(ORGANISATIONS_API + '/' + id)
      .then(() => {
        dispatch(resetOrganisations());
        dispatch(addSuccessNotification('Organisation Deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(error.message));
      });
  };
};

export const setSelectedOrganisation = (id) => ({
  type: SET_SELECTED_ORGANISATION,
  payload: id,
});

export const loadingOrganisations = () => ({
  type: SET_ORGANISATIONS_LOADING,
  payload: true,
});

export const stopOrganisationsLoading = () => ({
  type: SET_ORGANISATIONS_LOADING,
  payload: false,
});

export const getOrganisationByID = (data) => ({
  type: ADD_ORGANISATION,
  payload: data,
});

export const addOrganisationsList = (data, id) => (dispatch) => {
  const medium = getValues(data, 'medium');
  dispatch(addMediaList(medium));
  const applications = getValues(data, 'applications');
  dispatch(addApplicationList(applications));
  data.forEach((organisation) => {
    let users = [];
    organisation.roles = {};
    organisation.organisation_users.map((item) => {
      users.push(item.user);
      if (item.user.id === id) {
        organisation.role = item.role;
      }
      organisation.roles[item.user.id] = item.role;
      return null;
    });
    dispatch(addUsersList(users));
    organisation.users = getIds(users);
    organisation.applications = getIds(organisation.applications);
    deleteKeys([organisation], ['organisation_users', 'medium']);
  });
  const ids = getIds(data);
  dispatch({
    type: ADD_ORGANISATIONS,
    payload: {
      data: buildObjectOfItems(data),
      ids: ids,
    },
  });
};

export const resetOrganisations = () => ({
  type: RESET_ORGANISATIONS,
});

export const addOrganisationIds = (ids) => (dispatch) => {
  dispatch({
    type: ADD_ORGANISATION_IDS,
    payload: ids,
  });
};
