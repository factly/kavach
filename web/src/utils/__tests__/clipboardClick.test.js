import React from 'react';
import { useDispatch } from 'react-redux';
import { render, screen, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow, mount } from 'enzyme';
import ClipBoardCopy from '../clipboardClick';
import { act } from 'react-dom/test-utils';
import { addErrorNotification } from '../../actions/notifications';
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

const store = mockStore({});

describe('ClipBoardCopy', () => {
  it('should render the component', () => {
    let component = shallow(
      <Provider store={store}>
        <ClipBoardCopy text="my_token" />
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });
  it('should have a button with text "Copy"', () => {
    render(<ClipBoardCopy text="my_token" />);
    const button = screen.getByText('Copy');
    expect(button).toBeInTheDocument();
  });
  it('copies the token to the clipboard when "Copy" button is clicked', async () => {
    const writeTextMock = jest.fn();
    Object.defineProperty(navigator, 'clipboard', {
      value: {
        writeText: writeTextMock,
      },
    });

    jest.useFakeTimers();

    render(
      <Provider store={store}>
        <ClipBoardCopy text="my_token" />
      </Provider>,
    );
    act(() => {
      fireEvent.click(screen.getByText('Copy'));
    });

    // Simulate the passage of time until 5 seconds have passed
    jest.advanceTimersByTime(5000);

    expect(writeTextMock).toHaveBeenCalledWith('my_token');
  });

  it('displays an error message if the token could not be copied', async () => {
    const execCommandMock = jest.fn(() => false);
    Object.defineProperty(document, 'execCommand', {
      value: execCommandMock,
    });

    render(
      <Provider store={store}>
        <ClipBoardCopy text="my_token" />
      </Provider>,
    );
    act(() => {
      fireEvent.click(screen.getByText('Copy'));
    });
  });
});
