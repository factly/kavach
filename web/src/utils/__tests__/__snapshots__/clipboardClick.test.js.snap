// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClipBoardCopy should render the component 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ClipBoardCopy
    text="my_token"
  />
</ContextProvider>
`;
